### ISSUE TITLE
CoqIDE becomes unresponsive

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#4369]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 4369
Date: 2015-10-15 15:11:39 +0200
From: @_MSoegtropIMC
To:   
CC:   @_andrew-appel, coq-bugs-redist@lists.gforge.inria.fr, @_gares, kzi@cs.uni.wroc.pl, pauliankline@gmail.com, @_ppedrot, ranjani141@gmail.com, rogerwite@yahoo.co.uk

Last updated: 2016-08-30 21:10:49 +0200

## Bugzilla Comment ID: 11045
Date: 2015-10-15 15:11:39 +0200
From: @_MSoegtropIMC

Dear Coq Team,

I tried the 8.5-beta2 daily build from October 12th downloaded from  https://ci.inria.fr/coq/job/coq-8.5-win32/.

Unlike previosu versions, CoqIDE tends to become unresponsive regularly. One way (I am sure there are others) to reproduce the issue:

- Open CoqIDE
- Open any .v file
- Click the close box (header X button)

The window does not close and no further mouse clicks are accepted.

Best regards,

Michael

## Bugzilla Comment ID: 11295
Date: 2015-11-23 08:21:47 +0100
From: @_silene

*** Bug #4424 has been marked as a duplicate of this bug. ***

## Bugzilla Comment ID: 11624
Date: 2016-01-20 16:57:14 +0100
From: @_MSoegtropIMC

Dear Coq Team,

the issue is the same in 8.5rc1. Here are a few more sequences which reproducibly work or hang:

Here are a few UI sequences which work or hang reproducibly:

1. Open CoqIDE
2. Close via closebox => ok (3 out of 3)

1. Open CoqIDE
2. Open File
3. Close via closebox => HANG (3 out of 3)

1. Open CoqIDE
2. Open File
3. Close buffer => ok
4. Close via closebox ok (3 out of 3)

1. Open CoqIDE
2. Open File A
3. Open File B
4. Close buffer B => ok
5. Close buffer A => ok
6. Close via closebox ok (3 out of 3)

1. Open CoqIDE
2. Open File A
3. Open File B
4. Switch to File A buffer by clicking on tab 5. Close buffer A => HANG (3 out of 3)

With HANG I mean the UI becomes unresponsive and all I can do is kill CoqIDE.

There are quite a few other sequences which lead to a hang, also not involving close or close buffer, but as far as I remember all happened shortly after switching buffers via the tab. For these other use cases I couldn't find reproducible sequences as yet. But I hope the above sequences are sufficient to locate the bug.

Best regards,

Michael

## Bugzilla Comment ID: 11943
Date: 2016-02-19 17:04:31 +0100
From: @_ppedrot

*** Bug #4581 has been marked as a duplicate of this bug. ***

## Bugzilla Comment ID: 11945
Date: 2016-02-19 17:04:39 +0100
From: @_ppedrot

*** Bug #4578 has been marked as a duplicate of this bug. ***

## Bugzilla Comment ID: 11948
Date: 2016-02-19 17:28:04 +0100
From: @_gares

Copying here a test I failed to make "next week" but that is worth trying to progress on this bug.

> I can reproduce it, but only on windows.
> Can you try to comment out
>       ignore(CoqTop.wait h.proc);
> in ide/coq.ml, line 419?
> In theory this wait is performed only after having sent a message
> requesting Coq to quit, but something must have gone wrong...
> And I'm afraid that if Coq is not quitting, then even if CoqIDE closes a
> zombie process will stay around.

## Bugzilla Comment ID: 11952
Date: 2016-02-20 03:52:52 +0100
From: ranjani141@gmail.com

Hi,

Has this been resolved? It still seems to be there in coq 8.5 IDE

## Bugzilla Comment ID: 11966
Date: 2016-02-22 16:10:46 +0100
From: @_gares

Yes, it is still there and the fix I've proposed does not work :-/

## Bugzilla Comment ID: 12066
Date: 2016-03-09 16:33:34 +0100
From: @_gares

(In reply to Enrico Tassi from comment #7)
> Yes, it is still there and the fix I've proposed does not work :-/

No, it does!  I did it wrong, i.e. I commented out the kill, not the wait ;-)
I don't see zombies either... fingers crossed.

The bug seems yet another instance of OCaml-threads+sockets not working well on windows: before sending a kill -9, an OOB message (i.e. a message on the control channel) asks a "spawned coq process" to die.  The control thread blocked on a read is never woken up, hence nobody calls "exit 0" on windows ("kill -9" has no effect either).

If zombie coq processes are found, a more drastic option is to take the ide/ide_win32_stubs.c file and move it to lib/.  Such file provides a sort of kill for win32 playing with "consoles" and the like, but is quite scary to me.

## Bugzilla Comment ID: 13656
Date: 2016-08-30 21:10:49 +0200
From: @_ppedrot

*** Bug #4539 has been marked as a duplicate of this bug. ***
